src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.FunctionalTest/tests/Hover2_2.test.ts

```diff
@@ -0,0 +1,158 @@
+        await pollUntil(async () => {
+            await vscode.commands.executeCommand('workbench.action.closeAllEditors');
+            if (vscode.window.visibleTextEditors.length === 0) {
+                return true;
+            }
+
+            return false;
+        }, /* timeout */ 3000, /* pollInterval */ 500, true /* suppress timeout */);
+    });
+
+    test('Hovers over tags with multiple possible TagHelpers should return both', async () => {
+        const firstLine = new vscode.Position(0, 0);
+        await editor.edit(edit => edit.insert(firstLine, '<environment exclude="d" />\n'));
+        await editor.edit(edit => edit.insert(firstLine, '@addTagHelper *, SimpleMvc22\n'));
+        let hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(1, 3));
+
+        assert.ok(hoverResult, 'Should have returned a result');
+        assert.equal(hoverResult!.length, 1, 'Should only have one hover result since the markdown is presented as one.');
+        let mdString = hoverResult![0].contents[0] as vscode.MarkdownString;
+        assert.ok(mdString.value.includes('elements that conditionally renders'));
+        assert.ok(mdString.value.includes('I made it!'));
+
+        hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(1, 15));
+
+        assert.ok(hoverResult, 'Should have returned a result');
+        assert.equal(hoverResult!.length, 1, 'Should have a hover result for both EnvironmentTagHelpers');
+        mdString = hoverResult![0].contents[0] as vscode.MarkdownString;
+        assert.ok(mdString.value.includes('A comma separated list of environment names in'));
+        assert.ok(mdString.value.includes('Exclude it!'));
+    });
+
+    test('Can perform hovers on TagHelpers that need attributes', async () => {
+        const firstLine = new vscode.Position(0, 0);
+        await editor.edit(edit => edit.insert(firstLine, '<input class="someName" />\n'));
+        let hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(0, 3));
+
+        assert.ok(hoverResult, 'Should have returned a result');
+        assert.equal(hoverResult!.length, 1, 'Should not have a hover result for InputTagHelper because it does not have the correct attrs yet.');
+
+        await editor.edit(edit => edit.insert(firstLine, '<input asp-for="D" class="someName" />\n'));
+        hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(0, 3));
+
+        assert.ok(hoverResult, 'Should have a hover result for InputTagHelper.');
+        if (hoverResult) {
```

Comment:
lol isn't TypeScript the worst sometimes? Instead of wrapping all of this in the curly braces i'd isolate this into a single statement to keep things clearer. Recommendation:
```TypeScript
if (!hoverResult) {
    // Not possible, but strict TypeScript doesn't know about assert.ok above.
    return;
}
```

Same elsewhere.